consensus/system_contract/consensus.go

```diff
@@ -0,0 +1,390 @@
+	errInvalidUncleHash = errors.New("non empty uncle hash")
+	// errInvalidDifficulty is returned if a difficulty value is non-zero
+	errInvalidDifficulty = errors.New("non-one difficulty")
+	// errInvalidTimestamp is returned if the timestamp of a block is lower than
+	// the previous block's timestamp + the minimum block period.
+	errInvalidTimestamp = errors.New("invalid timestamp")
+	// errInvalidExtra is returned if the extra data is not empty
+	errInvalidExtra = errors.New("invalid extra")
+)
+
+// ErrUnauthorizedSigner is returned if a header is signed by a non-authorized entity.
+var ErrUnauthorizedSigner = errors.New("unauthorized signer")
+
+// SignerFn hashes and signs the data to be signed by a backing account.
+type SignerFn func(signer accounts.Account, mimeType string, message []byte) ([]byte, error)
+
+// Author implements consensus.Engine, returning the Ethereum address recovered
+// from the signature in the header's extra-data section.
+func (s *SystemContract) Author(header *types.Header) (common.Address, error) {
+	return ecrecover(header)
+}
+
+// VerifyHeader checks whether a header conforms to the consensus rules of a
+// given engine.
+func (s *SystemContract) VerifyHeader(chain consensus.ChainHeaderReader, header *types.Header, seal bool) error {
+	return s.verifyHeader(chain, header, nil)
+}
+
+// VerifyHeaders is similar to VerifyHeader, but verifies a batch of headers
+// concurrently. The method returns a quit channel to abort the operations and
+// a results channel to retrieve the async verifications (the order is that of
+// the input slice).
+func (s *SystemContract) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error) {
+	abort := make(chan struct{})
+	results := make(chan error, len(headers))
+
+	go func() {
+		for i, header := range headers {
+			err := s.verifyHeader(chain, header, headers[:i])
+			if err != nil {
+				log.Error("Error verifying headers", "err", err)
+			}
+			select {
+			case <-abort:
+				return
+			case results <- err:
+			}
+		}
+	}()
+	return abort, results
+}
+
+// verifyHeader checks whether a header conforms to the consensus rules.The
+// caller may optionally pass in a batch of parents (ascending order) to avoid
+// looking those up from the database. This is useful for concurrently verifying
+// a batch of new headers.
+func (s *SystemContract) verifyHeader(chain consensus.ChainHeaderReader, header *types.Header, parents []*types.Header) error {
+	if header.Number == nil {
+		return errUnknownBlock
+	}
+
+	// Don't waste time checking blocks from the future
+	if header.Time > uint64(time.Now().Unix()) {
+		return consensus.ErrFutureBlock
+	}
+	// Ensure that the nonce is zero
+	if header.Nonce != (types.BlockNonce{}) {
+		return errInvalidNonce
+	}
+	// Check that the extra-data contains signature
+	if header.Number.Cmp(big.NewInt(0)) != 0 && len(header.BlockSignature) != extraSeal {
```

Comment:
While we don't verify the signature for requested blocks, we still require them to have one?

Reply:
I thought this was the behavior, but updated now.